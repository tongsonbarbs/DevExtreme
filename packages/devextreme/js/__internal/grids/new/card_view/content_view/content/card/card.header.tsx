import type { dxToolbarItem } from '@js/ui/toolbar';
import { PureComponent } from '@ts/grids/new/grid_core/core/pure_component';
import { Button } from '@ts/grids/new/grid_core/inferno_wrappers/button';
import { CheckBox } from '@ts/grids/new/grid_core/inferno_wrappers/checkbox';
import type { InfernoNode } from 'inferno';

export const CLASSES = {
  cardHeader: 'dx-cardheader',
  cardHeaderItem: 'dx-cardheader-item',
  cardHeaderBefore: 'dx-cardheader-before',
  cardHeaderAfter: 'dx-cardheader-after',
};

export interface CardHeaderItem {
  location: 'before' | 'after';
  widget?: string;
  text?: string;
  options?: dxToolbarItem;
}

export interface CardHeaderProps {
  items: CardHeaderItem[];
}

export class CardHeader extends PureComponent<CardHeaderProps> {
  static defaultProps = {
    items: [] as CardHeaderItem[],
  };

  render(): InfernoNode {
    const renderWidget = (item: CardHeaderItem): InfernoNode => {
      switch (item.widget) {
        case 'dxCheckBox':
          return <CheckBox {...item.options} />;
        case 'dxButton':
          return <Button {...item.options} />;
        default:
          return <span>{item.text}</span>;
      }
    };

    return (
      <div className={CLASSES.cardHeader}>
        {this.props.items.map((item, index) => (
          <div
            // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
            className={`${CLASSES.cardHeaderItem} ${CLASSES.cardHeader}-${item.widget?.toLocaleLowerCase() || 'text'} ${
              item.location === 'before' ? CLASSES.cardHeaderBefore : CLASSES.cardHeaderAfter
            }`}
            key={index}
          >
            {renderWidget(item)}
          </div>
        ))}
      </div>
    );
  }
}
