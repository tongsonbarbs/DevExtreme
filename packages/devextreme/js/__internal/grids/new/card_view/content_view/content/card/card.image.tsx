import { PureComponent } from '@ts/grids/new/grid_core/core/pure_component';
import type { InfernoNode } from 'inferno';

const CLASSES = {
  cover: 'dx-card-cover',
};

export interface ImageProps {
  src?: string;
  alt?: string;
  style?: React.CSSProperties;
}

export class Image extends PureComponent<ImageProps> {
  render(): InfernoNode {
    if (!this.props.src) {
      return null;
    }
    return (
      <div className={CLASSES.cover}>
        <img
          src={this.props.src}
          alt={this.props.alt ?? 'Image'}
          style={{ width: '100%', height: 'auto' }}
        />
      </div>
    );
  }
}
