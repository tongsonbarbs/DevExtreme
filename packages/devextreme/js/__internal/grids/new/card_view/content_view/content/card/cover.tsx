/* eslint-disable @typescript-eslint/prefer-nullish-coalescing */
import { PureComponent } from '@ts/grids/new/grid_core/core/pure_component';
import type { InfernoNode } from 'inferno';

const CLASSES = {
  cover: 'dx-card-cover',
  image: 'dx-card-cover-image',
};

export interface CoverProps {
  src: string;
  alt?: string;
  className?: string;
}

export class Cover extends PureComponent<CoverProps> {
  render(): InfernoNode {
    const coverClasses = `${CLASSES.image} ${this.props.className || ''}`.trim();
    return (
      <div className={CLASSES.cover}>
        <img
          src={this.props.src}
          alt={this.props.alt ?? 'Devexpress'}
          className={coverClasses}
          // TODO: move to scss
          style={{ width: '100%', height: 'auto' }}
        />
      </div>
    );
  }
}
