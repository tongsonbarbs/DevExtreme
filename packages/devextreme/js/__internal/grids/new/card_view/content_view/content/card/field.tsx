/* eslint-disable @typescript-eslint/prefer-nullish-coalescing */
import { PureComponent } from '@ts/grids/new/grid_core/core/pure_component';
import type { InfernoNode, RefObject } from 'inferno';
import { createRef } from 'inferno';

export const CLASSES = {
  field: 'dx-cardview-field',
  fieldName: 'dx-cardview-field-name',
  fieldValue: 'dx-cardview-field-value',
  overflowHint: 'dx-cardview-overflow-hint',
};

export interface FieldProps {
  title: string | undefined;
  value: unknown;
  alignment: 'right' | 'center' | 'left';
  wordWrapEnabled?: boolean;
  cellHintEnabled?: boolean;
  elementRef?: RefObject<HTMLDivElement>;
  fieldTemplate?: (props: FieldProps) => JSX.Element;
  fieldCaptionTemplate?: (title: string) => JSX.Element;
  fieldValueTemplate?: (value: unknown) => JSX.Element;

  onFieldClick?: (e: MouseEvent) => void;
  onFieldDblClick?: (e: MouseEvent) => void;
  onFieldHoverChanged?: (hovered: boolean) => void;
  onFieldPrepared?: (element: HTMLElement) => void;
}

export class Field extends PureComponent<FieldProps> {
  private readonly containerRef: RefObject<HTMLDivElement>;

  constructor(props: FieldProps) {
    super(props);
    this.containerRef = this.props.elementRef || createRef<HTMLDivElement>();
  }

  componentDidMount(): void {
    const { onFieldPrepared } = this.props;
    if (onFieldPrepared && this.containerRef.current) {
      onFieldPrepared(this.containerRef.current);
    }
  }

  handleMouseEnter = (): void => {
    const { onFieldHoverChanged } = this.props;
    if (onFieldHoverChanged) {
      onFieldHoverChanged(true);
    }
  };

  handleMouseLeave = (): void => {
    const { onFieldHoverChanged } = this.props;
    if (onFieldHoverChanged) {
      onFieldHoverChanged(false);
    }
  };

  handleClick = (event: MouseEvent): void => {
    const { onFieldClick } = this.props;
    if (onFieldClick) {
      onFieldClick(event);
    }
  };

  handleDblClick = (event: MouseEvent): void => {
    const { onFieldDblClick } = this.props;
    if (onFieldDblClick) {
      onFieldDblClick(event);
    }
  };

  renderCaption(): JSX.Element | string {
    const { title, fieldCaptionTemplate } = this.props;
    if (fieldCaptionTemplate && title) {
      return fieldCaptionTemplate(title);
    }
    return <span className={CLASSES.fieldName}>{title}:</span>;
  }

  renderValue(): JSX.Element | string {
    const {
      value, fieldValueTemplate, wordWrapEnabled, alignment, cellHintEnabled,
    } = this.props;

    if (fieldValueTemplate && value) {
      return fieldValueTemplate(value);
    }

    const valueStyle = {
      textAlign: alignment,
      whiteSpace: wordWrapEnabled ? 'normal' : 'nowrap',
      overflow: 'hidden',
      textOverflow: 'ellipsis',
    };

    return (
      <span
        className={CLASSES.fieldValue}
        style={valueStyle}
        title={cellHintEnabled && typeof value === 'string' ? value : undefined}
      >
        {value}
      </span>
    );
  }

  render(): InfernoNode {
    const { fieldTemplate } = this.props;

    const containerProps = {
      onMouseEnter: this.handleMouseEnter,
      onMouseLeave: this.handleMouseLeave,
      onClick: this.handleClick,
      onDblClick: this.handleDblClick,
      ref: this.containerRef,
      tabIndex: 0,
      className: `${CLASSES.field}`.trim(),
    };

    if (fieldTemplate) {
      return fieldTemplate(this.props);
    }

    return (
      <div {...containerProps}>
        {this.renderCaption()}
        {this.renderValue()}
      </div>
    );
  }
}
