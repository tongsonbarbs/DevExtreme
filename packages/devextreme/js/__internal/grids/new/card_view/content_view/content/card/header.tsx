import type { dxToolbarItem } from '@js/ui/toolbar';
import { Toolbar } from '@ts/grids/new/grid_core/inferno_wrappers/toolbar';

export const CLASSES = {
  cardHeader: 'dx-cardheader',
  cardHeaderItem: 'dx-cardheader-item',
  cardHeaderBefore: 'dx-cardheader-before',
  cardHeaderAfter: 'dx-cardheader-after',
};

export interface CardHeaderItem {
  location: 'before' | 'after';
  widget?: string;
  text?: string;
  options?: dxToolbarItem;
}

export interface CardHeaderProps {
  items: CardHeaderItem[];
}

export const CardHeader = ({ items }: CardHeaderProps): JSX.Element => (
  <Toolbar items={items} />
);
