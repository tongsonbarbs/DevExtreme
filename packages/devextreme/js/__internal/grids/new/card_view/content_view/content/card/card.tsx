/* eslint-disable @typescript-eslint/prefer-nullish-coalescing */
import $ from '@js/core/renderer';
import type { DataRow } from '@ts/grids/new/grid_core/columns_controller/types';
import { PureComponent } from '@ts/grids/new/grid_core/core/pure_component';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { type ComponentType, type InfernoNode, render } from 'inferno';

import type { CoverProps } from './cover';
import { Cover } from './cover';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import type { FieldProps } from './field';
import { Field } from './field';
import type { CardHeaderItem } from './header';
import { CardHeader } from './header';

export const CLASSES = {
  card: 'dx-cardview-card',
};

export interface CardProps {
  row: DataRow;

  toolbar?: CardHeaderItem[];

  cover?: CoverProps;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  fieldTemplate?: any;
}

export class Card extends PureComponent<CardProps> {
  render(): InfernoNode {
    const FieldTemplate = this.props.fieldTemplate ?? Field;
    return (
      <div className={CLASSES.card} tabIndex={0}>
        <CardHeader
          items={this.props.toolbar || []}
        />
        {this.props.cover?.src && (
          <Cover
            src={this.props.cover.src}
            alt={this.props.cover.alt}
            className={this.props.cover.className}
          />
        )}
        {this.props.row.cells.map((cell, index) => (
          <FieldTemplate
            index={index}
            // eslint-disable-next-line max-len, @typescript-eslint/explicit-function-return-type
            defaultTemplate={{ render(model, _index, container) { render(<Field {...model} />, $(container).get(0)); } }}
            alignment={cell.column.alignment}
            title={cell.column.caption || cell.column.name}
            value={cell.value}
          />
        ))}
      </div>
    );
  }
}
