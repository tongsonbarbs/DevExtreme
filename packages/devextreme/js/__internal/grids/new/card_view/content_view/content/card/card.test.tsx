/* eslint-disable @typescript-eslint/no-unused-vars */
import { describe, expect, it } from '@jest/globals';
import { render } from 'inferno';

import { Card } from './card';

const props = {
  row: {
    cells: [
      {
        column: {
          dataField: 'Name',
          name: 'Field',
        },
        value: 'devextreme',
      },
    ],
    key: 0,
  },
  toolbar: [
    {
      location: 'before',
      widget: 'dxCheckBox',
    },
    {
      location: 'before',
      text: 'Card Header',
    },
    {
      location: 'after',
      widget: 'dxButton',
      options: {
        icon: 'edit',
        stylingMode: 'text',
      },
    },
    {
      location: 'after',
      widget: 'dxButton',
      options: {
        icon: 'trash',
        stylingMode: 'text',
      },
    },
  ],
  cover: {
    src: 'https://www.devexpress.com/support/demos/i/demo-thumbs/aspnetcore-grid.png',
    alt: 'Card Cover',
    className: 'cover-image',
  },
};

describe('Rendering', () => {
  it('should be rendered correctly', () => {
    const container = document.createElement('div');
    // @ts-expect-error
    render(<Card {...props} />, container);

    expect(container).toMatchSnapshot();
  });

  it('should render content correctly', () => {
    const container = document.createElement('div');
    // @ts-expect-error
    render(<Card {...props} />, container);

    const fieldValue = container.querySelector('.dx-cardview-field-value');
    expect(fieldValue?.textContent).toEqual('devextreme');
  });
});

describe('Card Header', () => {
  it('should render the card header components correctly', () => {
    const container = document.createElement('div');
    // @ts-expect-error
    render(<Card {...props} />, container);

    const cardHeaderText = container.querySelector('.dx-cardheader-text span');
    expect(cardHeaderText?.textContent).toBe('Card Header');

    const checkbox = container.querySelector('.dx-checkbox');
    expect(checkbox).not.toBeNull();

    const editButton = container.querySelector('.dx-icon-edit');
    expect(editButton).not.toBeNull();

    const trashButton = container.querySelector('.dx-icon-trash');
    expect(trashButton).not.toBeNull();
  });
});

describe('Image', () => {
  it('should render the image correctly', () => {
    const container = document.createElement('div');
    // @ts-expect-error
    render(<Card {...props} />, container);

    const image = container.querySelector('img');
    expect(image).not.toBeNull();
    expect(image?.src).toBe(props.cover.src);
    expect(image?.alt).toBe(props.cover.alt);
  });
});

describe('Field Template', () => {
  it('should render field template correctly', () => {
    const container = document.createElement('div');
    // @ts-expect-error
    render(<Card {...props} />, container);

    const fieldName = container.querySelector('.dx-cardview-field-name');
    const fieldValue = container.querySelector('.dx-cardview-field-value');

    expect(fieldName?.textContent).toBe('Field: ');
    expect(fieldValue?.textContent).toBe('devextreme');
  });
});
