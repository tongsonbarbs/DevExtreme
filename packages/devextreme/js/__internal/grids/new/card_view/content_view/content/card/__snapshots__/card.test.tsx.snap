// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering should be rendered correctly 1`] = `
<div>
  <div
    class="dx-cardview-card"
    tabindex="0"
  >
    <div
      class="dx-cardheader"
    >
      <div
        class="dx-cardheader-item dx-cardheader-dxcheckbox dx-cardheader-before"
      >
        <div
          aria-checked="false"
          aria-invalid="false"
          aria-readonly="false"
          class="dx-widget dx-checkbox"
          role="checkbox"
          tabindex="0"
        >
          <input
            type="hidden"
            value="false"
          />
          <div
            class="dx-checkbox-container"
          >
            <span
              class="dx-checkbox-icon"
            />
            
          </div>
        </div>
      </div>
      <div
        class="dx-cardheader-item dx-cardheader-text dx-cardheader-before"
      >
        <span>
          Card Header
        </span>
      </div>
      <div
        class="dx-cardheader-item dx-cardheader-dxbutton dx-cardheader-after"
      >
        <div
          aria-label="edit"
          class="dx-widget dx-button dx-button-mode-text dx-button-normal dx-button-has-icon"
          role="button"
          tabindex="0"
        >
          <div
            class="dx-button-content"
          >
            <i
              class="dx-icon dx-icon-edit"
            />
          </div>
        </div>
      </div>
      <div
        class="dx-cardheader-item dx-cardheader-dxbutton dx-cardheader-after"
      >
        <div
          aria-label="trash"
          class="dx-widget dx-button dx-button-mode-text dx-button-normal dx-button-has-icon"
          role="button"
          tabindex="0"
        >
          <div
            class="dx-button-content"
          >
            <i
              class="dx-icon dx-icon-trash"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="dx-card-cover"
    >
      <img
        alt="Card Cover"
        class="dx-card-cover-image cover-image"
        src="https://www.devexpress.com/support/demos/i/demo-thumbs/aspnetcore-grid.png"
        style="width: 100%; height: auto;"
      />
    </div>
    <div
      class="dx-cardview-field"
      tabindex="0"
    >
      <span
        class="dx-cardview-field-name"
      >
        Field
        : 
      </span>
      <span
        class="dx-cardview-field-value"
      >
        devextreme
      </span>
    </div>
  </div>
</div>
`;
